pkg/kn/commands/common/human_readable_flags.go

```diff
@@ -8,74 +8,62 @@
 //
 // Unless required by applicable law or agreed to in writing, software
 // distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or im
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package commands
+package common
 
 import (
 	"fmt"
+	"time"
+
 	hprinters "github.com/knative/client/pkg/printers"
 	"github.com/knative/pkg/apis"
 	duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+	serving "github.com/knative/serving/pkg/apis/serving"
 	servingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
 	"github.com/spf13/cobra"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	metav1beta1 "k8s.io/apimachinery/pkg/apis/meta/v1beta1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/duration"
-	"k8s.io/cli-runtime/pkg/genericclioptions"
-	"time"
 )
 
-// ServiceGetFlags composes common printer flag structs
-// used in the Get command.
-type ServiceGetFlags struct {
-	GenericPrintFlags  *genericclioptions.PrintFlags
-	HumanReadableFlags *HumanPrintFlags
+// HumanPrintFlags provides default flags necessary for printing.
+// Given the following flag values, a printer can be requested that knows
+// how to handle printing based on these values.
+type HumanPrintFlags struct {
+	//TODO: Add more flags as required
 }
 
-// AllowedFormats is the list of formats in which data can be displayed
-func (f *ServiceGetFlags) AllowedFormats() []string {
-	formats := f.GenericPrintFlags.AllowedFormats()
-	formats = append(formats, f.HumanReadableFlags.AllowedFormats()...)
-	return formats
+// AllowedFormats returns more customized formating options
+func (f *HumanPrintFlags) AllowedFormats() []string {
+	// TODO: Add more formats eg: wide
+	return []string{""}
 }
 
-// ToPrinter attempts to find a composed set of ServiceGetFlags suitable for
-// returning a printer based on current flag values.
-func (f *ServiceGetFlags) ToPrinter() (hprinters.ResourcePrinter, error) {
-	// if there are flags specified for generic printing
-	if f.GenericPrintFlags.OutputFlagSpecified() {
-		p, err := f.GenericPrintFlags.ToPrinter()
-		if err != nil {
-			return nil, err
-		}
-		return p, nil
-	}
-	// if no flags specified, use the table printing
-	p, err := f.HumanReadableFlags.ToPrinter()
-	if err != nil {
-		return nil, err
-	}
+// ToPrinter receives returns a printer capable of
+// handling human-readable output.
+func (f *HumanPrintFlags) ToPrinter() (hprinters.ResourcePrinter, error) {
+	p := hprinters.NewTablePrinter(hprinters.PrintOptions{})
+	// Add the column definitions and respective printing functions for service get command
+	ServiceGetHandlers(p)
```

Comment:
The structure here feels wrong — I'd like to see the common flag stuff here in this file, but the resource-specific stuff being able to be dynamic per resource it is. Perhaps there's a function-typed member of the HumanPrintFlags struct that does that, that's set by each individual package, or...

Anyway, this needs some thinking.

Reply:
Yup, I can see that, let's address it after.

I re-arranged things a bit but still dealing with cyles in test. Because all tests call the `NewKnCommand` function and that is where all sub commands are added which means it imports all the subpackages. Hence the cycle in tests code :(

I just pushed my latest attempt. You can see there. I'll try to do the refactor since I am stuck on the tests code.

Reply:
Solved in last push